### ISSUE TITLE
Cannot load incompatible bytecode (compiling bytecode with luajit)

### ISSUE THREAD

**Issue by Source61** [edubart/otclient#1205]
### Steps to reproduce (include any configuration/script required to reproduce)
1. Compile otclient
2. Compile a lua module such as /modules/corelib/math.lua to bytecode using LATEST LuaJIT version
3. Run otclient and check otclient.log

### Expected behaviour
Should run without an issue

### Actual behaviour
Complains about incompatible bytecode and exits.

### Environment
Windows 10 with Debian 12 WSL.
Modified OTClient version from 2-3 years ago. Never touched src/framework/luaengine/*.
The issue seems to have started happening after I upgraded all the software on the WSL one day about 6-9 months ago and likely LuaJIT was upgraded as well, never had this issue before and been using LuaJIT compilation for years.
I've also tried upgrading to the most recent LuaJIT (and all other) vcpkg libraries to no avail.

This issue incidentally happened around the same time (9 months ago) https://github.com/neovim/neovim/issues/18169
https://stackoverflow.com/questions/72875661/i-got-this-error-after-upgrading-neovim-to-version-0-7-2-2

**Comment by diath** [comment]
The issue always stems from incompatible LuaJIT versions between the one you used to compile the Lua files to bytecode and the one you linked the client against, please keep in mind that LuaJIT doesn't guarantee bytecode compatibility even for MINOR releases.

**Comment by sonoro1234** [comment]
https://github.com/LuaJIT/LuaJIT/issues/967#issuecomment-1448580644

This comment should be here, not in LuaJIT. Just to discover: 
https://github.com/edubart/otclient/commit/e3caaacc55b68d27176645fbc2dc059398c9458f

**Comment by sonoro1234** [comment]
Also, if you close the issue it is very difficult to trace it.
Perhaps you should reopen it and reread diath's comment.

**Comment by Source61** [comment]
> The issue always stems from incompatible LuaJIT versions between the one you used to compile the Lua files to bytecode and the one you linked the client against, please keep in mind that LuaJIT doesn't guarantee bytecode compatibility even for MINOR releases.

Indeed. But the problem is, as I commented here today https://github.com/LuaJIT/LuaJIT/issues/967#issuecomment-1448580644, that updating otclient-sdk with the latest LuaJIT version didn't help as the application crashed (application cannot start or be opened or something like that), and I saw a warning while compiling otclient with one of the luaengine files about redeclaring macros, I don't know, I just think it might be an issue with otclient at this point since updating LuaJIT in vcpkg, in otclient-sdk, and the LuaJIT binary, did not help - in fact otclient crashes with the latest LuaJIT version.

I now use an old LuaJIT binary which works for me, but underlying the issue there seem to be an issue with otclient/outdated LuaJIT code as far as I can tell.

> Also, if you close the issue it is very difficult to trace it. Perhaps you should reopen it and reread diath's comment.

I didn't close any of the issues, they were closed by the maintainers. Thanks for commenting here. I don't feel it's appropriate to re-create a new issue when the first one was closed though.
I feel like if the maintainer want to close the issue, he wants to close the issue, he should re-open it if he understands the issue, if he doesn't understand or accept the issue then there's nothing I can do, spamming new issues won't help, that's how I feel.

**Comment by Source61** [comment]
> > that updating otclient-sdk with the latest LuaJIT version didn't help as the application crashed
> 
> The client crashes because it cannot load incompatible bytecode from corelib, and an error in corelib is considered a fatal error because it's essential for the client to function. 

No, it gives me an "incompatible bytecode" error in the logs when there's a version mismatch, this crashes the client before it even gets to open the log, before any Lua code is loaded, basically somewhere in the start of or pre-main.
If you want I can try to run the client in a debugger and see what line is causing the crash, if you're willing to look this through.

> > but underlying the issue there seem to be an issue with otclient/outdated LuaJIT code as far as I can tell.
> 
> Always compare the output of `luajit -v` in the terminal and `print(jit.version)` in the game client, if they don't match, that's an issue with using incompatible versions, not with otclient/LuaJIT, vcpkg doesn't always immediately update their packages to the newest versions, so if you download the latest LuaJIT from their website, and update vcpkg LuaJIT to latest, that doesn't automatically imply they're both the same version.

Hard to print(jit.version) in a client that won't allow you to load the Lua code necessary to run a print statement.
Not even that, but the client crashes way before Lua code is even run, so I can't just temporarily avoid loading bytecode as a temporary bypass.

**Comment by diath** [comment]
That's unrelated to your issue with incompatible bytecode.
I can build and run the client just fine using a fresh Win10 VM and latest LuaJIT from vcpkg so it's unlikely that the issue is related to either otclient or LuaJIT. Considering that your client crashes before it even enters the main routine, it's probably related to your compiler setup/libs/DLL files.

**Comment by Source61** [comment]
> That's unrelated to your issue with incompatible bytecode. I can build and run the client just fine using a fresh Win10 VM and latest LuaJIT from vcpkg so it's unlikely that the issue is related to either otclient or LuaJIT. Considering that your client crashes before it even enters the main routine, it's probably related to your compiler setup/libs/DLL files.

Well the question is what LuaJIT cli version you're using to compile the bytecode.
It worked fine for me as well using Debian 11, then I upgraded to Debian 12 (Testing) and it still was fine, but then about 9-11 months ago they updated the version, so when I upgraded my LuaJIT cli version it broke compatibility with the client (otclient SDK).
It doesn't seem that the issue is vcpkg, but with otclient-SDK like I said. In fact print(jit.version) does NOT print my vcpkg version, it prints my otclient-SDK version, which is LuaJIT 2.1.0-beta1.
Removing otclient-SDK doesn't work as it will complain about missing LuaJIT, at least I'm pretty sure, as I've tried to remove other otclient-SDK libraries and have it complain to me about this.
